datastore/store.go

```diff
@@ -197,72 +197,26 @@ func (db *CachedDB) VRFNonce(id types.NodeID, epoch types.EpochID) (types.VRFPos
 	return nonce, nil
 }
 
-// GetAtxHeader returns the ATX header by the given ID. This function is thread safe and will return an error if the ID
+// GetAtx returns the ATX by the given ID. This function is thread safe and will return an error if the ID
 // is not found in the ATX DB.
-func (db *CachedDB) GetAtxHeader(id types.ATXID) (*types.ActivationTxHeader, error) {
+func (db *CachedDB) GetAtx(id types.ATXID) (*types.ActivationTx, error) {
 	if id == types.EmptyATXID {
 		return nil, errors.New("trying to fetch empty atx id")
 	}
 
-	if atxHeader, gotIt := db.atxHdrCache.Get(id); gotIt {
-		return atxHeader, nil
-	}
-
-	return db.getAndCacheHeader(id)
-}
-
-// GetFullAtx returns the full atx struct of the given atxId id, it returns an error if the full atx cannot be found
-// in all databases.
-func (db *CachedDB) GetFullAtx(id types.ATXID) (*types.ActivationTx, error) {
-	if id == types.EmptyATXID {
-		return nil, errors.New("trying to fetch empty atx id")
+	if atx, gotIt := db.atxCache.Get(id); gotIt {
+		return atx, nil
 	}
 
 	atx, err := atxs.Get(db, id)
 	if err != nil {
-		return nil, fmt.Errorf("get ATXs from DB: %w", err)
+		return nil, fmt.Errorf("get ATX from DB: %w", err)
 	}
 
-	db.atxHdrCache.Add(id, atx.ToHeader())
+	db.atxCache.Add(id, atx)
 	return atx, nil
 }
 
-// getAndCacheHeader fetches the full atx struct from the database, caches it and returns the cached header.
-func (db *CachedDB) getAndCacheHeader(id types.ATXID) (*types.ActivationTxHeader, error) {
-	_, err := db.GetFullAtx(id)
-	if err != nil {
-		return nil, err
-	}
-
-	atxHeader, gotIt := db.atxHdrCache.Get(id)
-	if !gotIt {
-		return nil, fmt.Errorf("inconsistent state: failed to get atx header: %w", err)
-	}
-
-	return atxHeader, nil
-}
-
-// IterateEpochATXHeaders iterates over ActivationTxs that target an epoch.
-func (db *CachedDB) IterateEpochATXHeaders(
-	epoch types.EpochID,
-	iter func(*types.ActivationTxHeader) error,
-) error {
-	ids, err := atxs.GetIDsByEpoch(context.Background(), db, epoch-1)
-	if err != nil {
-		return err
-	}
-	for _, id := range ids {
-		header, err := db.GetAtxHeader(id)
-		if err != nil {
-			return err
-		}
-		if err := iter(header); err != nil {
-			return err
-		}
-	}
-	return nil
-}
```

Comment:
As @ivan4th already mentioned, I agree we need this function as part of the `sql/atxs` package and taking a callback for `types.ActivationTx`. There is already `IterateAtxsData` which could be used instead of this method.

Reply:
Or `IterateAtxsOps`

Reply:
I inlined the query code where `CachedDB::IterateEpochATXHeaders` was used for two reasons:
- each call site has different needs:
  * they need a different small subset of fields,
  * some need to also query another table, for example see if ID is malicious.
- putting very specialized functions into the `sql/atxs` package felt artificial/wrong as they would only be used in a single place.

It feels inefficient to query all fields of an ATX with a generic method when only a few are needed. It's also problematic how to do subqueries with a generic method like the suggested `IterateAtxsOps`. Doing a subquery within the decoder could lead to deadlocks when all decoders cannot make progress waiting to obtain a DB connection and starving.

We already have some specialized code that interacts with the DB next to the logic that uses it and I think this is right.

Reply:
Moved the code into `sql/atxs`